src/service/layers-impl.js

```diff
@@ -0,0 +1,634 @@
+    this.parentLayer_ = undefined;
+
+    this.needsRemeasure_ = true;
+    this.size_ = SizeWh(0, 0);
+    this.position_ = PositionLt(0, 0);
+
+    this.isLayer_ = false;
+    this.isRootLayer_ = false;
+    this.needsScrollRemeasure_ = false;
+    this.scrollLeft_ = 0;
+    this.scrollTop_ = 0;
+
+    this.children_ = [];
+  }
+
+  /**
+   * Gets the element's LayoutElement instance.
+   * @param {!Element}
+   * @return {!LayoutElement}
+   */
+  static for(element) {
+    return dev().assert(LayoutElement.forOptional(element));
+  }
+
+  /**
+   * Gets the element's LayoutElement instance, if the element has one.
+   * @param {!Element}
+   * @return {?LayoutElement}
+   */
+  static forOptional(element) {
+    return element[LAYOUT_PROP];
+  }
+
+  /**
+   * Finds the element's parent layer
+   * If the element is itself a layer, it returns the layer's parent layer.
+   * @param {!Element}
+   * @param {opt_force=} Whether to force a re-lookup
+   * @return {?LayoutElement}
+   */
+  static getParentLayer(element, opt_force) {
+    if (!opt_force) {
+      const layout = LayoutElement.forOptional(element);
+      if (layout) {
+        return layout.getParentLayer();
+      }
+    }
+
+    const win = element.ownerDocument.defaultView;
+    if (computedStyle(win, element).position == 'fixed') {
+      LayoutLayers.declareLayer(element);
+      return null;
+    }
+
+    let last = element;
+    let op = element.offsetParent;
+    for (let el = last.parentNode; el; el = el.parentNode) {
+      const layout = LayoutElement.forOptional(el);
+      if (layout && layout.isLayer()) {
+        return layout;
+      }
+
+      // now check to see if offsetParent is a fixed layer
+      if (el === op) {
+        if (computedStyle(win, op).position == 'fixed') {
+          LayoutLayers.declareLayer(op);
+          return op;
+        }
+        op = op.offsetParent;
+      }
+
+      last = el;
+    }
+
+
+    if (last.nodeType !== Node.DOCUMENT_NODE) {
+      throw dev().createError('element is not in the DOM tree');
+    }
+    return null;
+  }
+
+  contains(element) {
+    const root = this.element_;
+    return root !== element && root.contains(element);
+  }
+
+  add(child) {
+    dev().assert(this.isLayer());
+    dev().assert(this.contains(child));
+
+    const layout = LayoutElement.for(child);
+    this.children_.push(layout);
+  }
+
+  remove(child) {
+    dev().assert(this.isLayer());
+
+    const layout = LayoutElement.for(child);
+    dev().assert(layout.getParentLayer() === this);
+
+    layout.undeclareLayer();
+    layout.forgetParentLayer();
+    const i = this.children_.indexOf(layout);
+    if (i > -1) {
+      this.children_.splice(i, 1);
+    }
+  }
+
+  isLayer() {
+    return this.isLayer_;
+  }
+
+  declareLayer() {
+    this.isLayer_ = true;
+    this.needsRemeasure_ = true;
+    this.needsScrollRemeasure_ = true;
+
+    const element = this.element_;
+    const scroller = getScrollingElement(element);
```

Comment:
The root scrolling layer is either to identify. Why not just call `document.documentElement` (or `ampdoc.rootElement`) a root scrollable layer once and not have to recalculate it here? It'd also be desirable to reconcile all weirdness about which element is considered document scroller in which browser by just making `ampdoc.rootElement` be that in all cases.

The bottom line, this implementation can be very simple: `if (element == this.ampdoc_.rootElement) {this-is-root-layer-element-for-ampdoc}`

Reply:
For simplicity, it would still be nice to say the root document layer is anchored at some very predictable element, such as rootElement. All scrolling features would come from Viewport for the root layer - ultimately it doesn't really matter what a platform considers to be scrollingElement. It only matters which element is the position reference point for other elements within. And we know that the browsers (slowly) moving toward this as well: `scrollingElement == rootElement == documentElement`

Reply:
In other words: rootElement doesn't need to be used to get scrollLeft/Top - layer could get this info independently.

Reply:
Unfortunately, we cannot just identify `documentElement` as the scrolling element because of how DOM treats the `scrollTop` and `getBoundingClientRect().y`.

For normal elements, when they scroll (so `scrollTop` increases), the `rect.y` is constant, which absolutely makes sense because the element's inner coords scrolled, not the element. But for root scrollers, the `rect.y === -scrollTop`, so the root scroller's position _changes_ just because it is scrolled. If I were to leave that uncorrected, all my APIs would double count the scroll and wreak havoc on intersection calculations (essentially, anything outside of the first viewport would never "enter" viewport).

So I have to know exactly which element has this behavior, and which doesn't. Hence the need for the Viewport to now instantiate Layers with the root scroller.

Reply:
Ok. I'll re-review this part again.